### ISSUE TITLE
low performance in semi-supervised mode when employing weakly_loss with 2 gpus

### ISSUE THREAD

**Issue by wqhIris** [yassouali/CCT#37]
Thank you for your nice work!

I tried to training the model with 1464 labeled samples in semi-supervised mode, and I used 2 gpus. I set the epoch as 80, and stop it after 50 epoch. But the performance is poor, e.g., miou at epoch 5 is 34.70% while at epoch 10 is 11.40%.
![image](https://user-images.githubusercontent.com/33059293/117651594-b139e580-b1c4-11eb-9ac0-0389aa5d38fc.png)

I set the 'use_weak_labels' as true, the 'drop_last' as false, and the rest are default.

Have you ever met this situation?

**Comment by yassouali** [comment]
Hi @wqhIris 

Did you check that the generated pseudo labels are correct? and also check again that the config used are the same as the one provided
Can you also maybe reduce the learning rate a little but since the val is increasing for you case

**Comment by wqhIris** [comment]
Hi @yassouali, thank you for rapid reply.

The generated pseudo labels are created by following the introductions in README.md, and I ever used these pseudo labels to train models with same settings, but use 1 gpu, at that time the miou is about 73%, so I think the pseudo labels are correct. 
But it is a little strange, when I use 2 gpu, the performance dropped.
 
I compared the config file with yours in the github, they are same except the 'use_weak_labels', the 'drop_last', some file path. The followings are my config setting
 {
    "name": "CCT",
    "experim_name": "CCT",
    "n_gpu": 2,
    "n_labeled_examples": 1464,
    "diff_lrs": true,
    "ramp_up": 0.1,  
    "unsupervised_w": 30,
    "ignore_index": 255,
    "lr_scheduler": "Poly",
    "use_weak_lables":true,
    "weakly_loss_w": 0.4,
    "pretrained": true,

    "model":{
        "supervised": false,
        "semi": true,
        "supervised_w": 1,

        "sup_loss": "CE",
        "un_loss": "MSE",

        "softmax_temp": 1,
        "aux_constraint": false,
        "aux_constraint_w": 1,
        "confidence_masking": false,
        "confidence_th": 0.5,

        "drop": 6,
        "drop_rate": 0.5,
        "spatial": true,    

        "cutout": 6,
        "erase": 0.4,
    
        "vat": 2,
        "xi": 1e-6,
        "eps": 2.0,

        "context_masking": 2,
        "object_masking": 2,
        
        "feature_drop": 6,

        "feature_noise": 6,
        "uniform_range": 0.3
    },


    "optimizer": {
        "type": "SGD",
        "args":{
            "lr": 1e-2,
            "weight_decay": 1e-4,
            "momentum": 0.9
        }
    },


    "train_supervised": {
        "data_dir": "/mnt/10102006/CCT/VOCtrainval_11-May-2012",
        "batch_size": 10,
        "crop_size": 320,
        "shuffle": true,
        "base_size": 400,
        "scale": true,
        "augment": true,
        "flip": true,
        "rotate": false,
        "blur": false,
        "split": "train_supervised",
        "num_workers": 8,
        "drop_last": false
    },

    "train_unsupervised": {
        "data_dir": "/mnt/10102006/CCT/VOCtrainval_11-May-2012",
        "weak_labels_output": "/mnt/10102006/CCT/pseudo_labels",
        "batch_size": 10,
        "crop_size": 320,
        "shuffle": true,
        "base_size": 400,
        "scale": true,
        "augment": true,
        "flip": true,
        "rotate": false,
        "blur": false,
        "split": "train_unsupervised",
        "num_workers": 8,
        "drop_last": false
    },

    "val_loader": {
        "data_dir": "/mnt/10102006/CCT/VOCtrainval_11-May-2012",
        "batch_size": 1,
        "val": true,
        "split": "val",
        "shuffle": false,
        "num_workers": 4,
        "drop_last": false
    },

    "trainer": {
        "epochs": 80,
        "save_dir": "saved_epoch80_1464label_CCTweakloss/",
        "save_period": 5,
  
        "monitor": "max Mean_IoU",
        "early_stop": 10,
        
        "tensorboardX": true,
        "log_dir": "saved_epoch80_1464label_CCTweakloss/",
        "log_per_iter": 20,

        "val": true,
        "val_per_epochs": 5
    }
}

In order to use 'drop_last', I add some modifications in dataloaders/voc.py and base/base_dataloader.py, I guess these modifications may not impact the performance a lot.
![image](https://user-images.githubusercontent.com/33059293/117791591-13562180-b27d-11eb-8e9f-07a60ed82444.png)
![image](https://user-images.githubusercontent.com/33059293/117791592-13562180-b27d-11eb-8749-bedbb782d65d.png)

By the way, as for decreasing the lr, I tried to change the learning rate to 0.001, but may be in tomorrow the training process will finish.

**Comment by wqhIris** [comment]
Hi, @yassouali

I changed the learning rate to 0.001, the best miou in 50 epochs (the total epoch is 80, but stop at No.50 epoch) is about 67%, but is still lower than the result in paper (73.4%). If I changed learning rate to 0.0001, the results dropped to 37.5%.

Should I stop at larger epoch when setting learning rate as 0.001, or adjust other parameters?

**Comment by yassouali** [comment]
@wqhIris 

In your case you'll need to spend a little time to find the best parameters, maybe try 0.005 and see, you can also reduce the number of aux. decoders for quick experimentation (maybe K =2 for all aux. decoder) and then go back after finding the correct lr.

**Comment by wqhIris** [comment]
Thank you a lot, I will try it again referring to your suggestions.
